Add "readonly" as a keyword for PHP and add previous classnames to descriptor pool

bshaffer opened the pull request:
See #10038 

This PR adds compatibility for PHP 8.1 by identifying "readonly" as a reserved word, and by adding forwards compatibility for any `ReadOnly` classes generated with a previous protobuf version (see new tests / new test protos).

It should be noted that in the past when we added a new keyword, forwards compatibility was not preserved. We only realized it now because we have several libraries (Spanner/Datastore/Firestore) which use the keyword `readonly` for message names. 

## Summary
- Makes `ReadOnly` a reserved keyword for PHP 8.1
    - Added in #9633
    - When the forwards-compatibility issue was discovered, we reverted this in #10036
- Adds the "previously unreserved name" to the descriptor pool to preserve forwards compatibility.
- Adds tests with previously generated classes to verify it works
- Updates the `protobuf.so` extension as well!

## BC file for the previously unreserved classnames

In the file `ReadOnly.php`, where there used to generate `class ReadOnly`, there is now the following code to create an alias for the new classname:
```php
<?php
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: proto/test_reserved_message_lower.proto

namespace Lower;

class_exists(PBReadOnly::class); // autoload the new class, which will also create an alias to the deprecated class
@trigger_error(__NAMESPACE__ . '\ReadOnly is deprecated and will be removed in the next major release. Use PBReadOnly instead', E_USER_DEPRECATED);
```

This ensures that anyone still using the name `Readme` on previous versions of PHP will not be broken. 

## Alias declaration in new class file to the previously unreserved classnames

Likewise, at the bottom of the new `PBReadOnly.php` file/class, we create an alias to the old one:

```php
/**
 * Generated from protobuf message <code>readonly</code>
 */
class PBReadOnly extends \Google\Protobuf\Internal\Message
{

    /**
     * Constructor.
     *
     * @param array $data {
     *     Optional. Data for populating the Message object.
     *
     * }
     */
    public function __construct($data = NULL) {
        \GPBMetadata\Proto\TestReservedMessageLower::initOnce();
        parent::__construct($data);
    }

}

// Adding a class alias for backwards compatibility with the "readonly" keyword.
class_alias(ReadOnly::class, __NAMESPACE__ . '\ReadOnly');
```

**NOTE**: The classname "readonly" is still valid, but the syntax for declaring it (e.g. `class ReadOnly {}`) is not. That means that creating an alias for it, and instantiating the class directly, is still valid in PHP 8.1.

## Previously unreserved classnames

The previously unreserved classnames function the same way as any other type of reserved word, with the exception that they also get added to the descriptor pool. This is to maintain backwards compatibility.

In order for Protobuf to be able to know and handle any messages defined using the previous classname (e.g. `ReadOnly`) for versions of any library that are still using the older versions, we've added the previous names to the descriptor pool in both the native PHP and the C-extension implementations.

haberman commented on php/ext/google/protobuf/protobuf.c:
```diff
@@ -242,13 +242,18 @@ bool ObjCache_Get(const void *upb_obj, zval *val) {
+  char *k = GetPhpClassname(upb_MessageDef_File(m), upb_MessageDef_FullName(m), false);
   zend_hash_str_add_ptr(&PROTOBUF_G(name_msg_cache), k, strlen(k), (void*)m);
+  char *k2 = GetPhpClassname(upb_MessageDef_File(m), upb_MessageDef_FullName(m), true);
+  if (strcmp(k, k2) != 0) {
```
We need to be careful about the CPU costs of what we are doing here. The cost of loading large schemas is significant, and this is in the critical path for important users like Ads.

As written this will always generate the name twice, compare them, and throw away one if they are the same. Could we change it to only generate the alternate name if the primary name is a reserved word?

bshaffer replied:
The problem is by the time we can check if the name is reserved, we're way deep in the `is_reserved` function, and then we lose the ability to compare once the full PHP classname is returned.

bshaffer replied:
I refactored it to do as I mention above (read the ending classname and return `true` if it was previously unreserved)

I think there's probably a better way and I'm definitely open to ideas.

bshaffer commented:
Looks like I need to add some of the new test files to the distribution file! Should be an easy fix.

bshaffer commented:
@haberman This is ready for another test run!

bshaffer commented:
@zeriyoshi I see the following errors:
```
-- Performing Test HAVE_CXX_FLAG_WD654 - Failed
-- Performing Test HAVE_THREAD_SAFETY_ATTRIBUTES -- failed to compile
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
CMake Error at CMakeLists.txt:34 (message):
  Did not find Google Test sources! Either pass correct path in
  GOOGLETEST_PATH, or enable BENCHMARK_DOWNLOAD_DEPENDENCIES, or disable
  BENCHMARK_USE_BUNDLED_GTEST, or disable BENCHMARK_ENABLE_GTEST_TESTS /
  BENCHMARK_ENABLE_TESTING.
```
Could these be what is causing the `make` error?

bshaffer commented:
@zeriyoshi okay I fixed it, for some reason in my commit there was an update to the `third_party/benchmarks` submodule. I don't know how that got in there, but I removed the change and that should do the trick!

fowles commented on src/google/protobuf/compiler/php/php_generator.cc:
```diff
@@ -420,6 +420,16 @@ std::string LegacyGeneratedClassFileName(const DescriptorType* desc,
 
+template <typename DescriptorType>
+std::string LegacyReadOnlyGeneratedClassFileName(const DescriptorType* desc,
+                                   const Options& options) {
```
formatting is off.  Please run clang-format

fowles commented on src/google/protobuf/compiler/php/php_generator.cc:
```diff
@@ -1539,6 +1588,19 @@ void GenerateMessageFile(const FileDescriptor* file, const Descriptor* message,
+        "new", fullname,
+        "old", message->name());
+    LegacyReadOnlyGenerateClassFile(file, message, options, generator_context);
+  }
```
this looks to duplicate the code above.  Can you add well named helper?

bshaffer commented on php/ext/google/protobuf/def.c:
```diff
@@ -500,14 +500,24 @@ static void Descriptor_destructor(zend_object* obj) {
+
+    if (!ce) {
+      zend_error(E_ERROR, "Couldn't load generated class %s", classname);
+    }
```
This crashes the program, I'm not sure why

bshaffer replied:
The problem goes away when I run `git -xdf ext` before running `composer test_c`. Apparently some files were being cached which needed to be cleared for this to work.

bshaffer replied:
done

bshaffer commented on php/ext/google/protobuf/names.c:
```diff
@@ -234,12 +252,43 @@ char *GetPhpClassname(const upb_FileDef *file, const char *fullname) {
+  }
+  classname += 2;
+  int length = strlen(classname);
+  char* lower = calloc(1, length + 1);
```
Changing this to `malloc` also crashes the program, and I'm not sure why:

```c
char* lower = malloc(length + 1);
int i = 0;
while(lower[i]) {
  lower[i] = nolocale_tolower(classname[i]);
  i++;
}
```

bshaffer commented on php/ext/google/protobuf/protobuf.c:
```diff
@@ -242,13 +242,18 @@ bool ObjCache_Get(const void *upb_obj, zval *val) {
+    char *k2 = GetPhpClassname(upb_MessageDef_File(m), upb_MessageDef_FullName(m), true);
+    zend_hash_str_add_ptr(&PROTOBUF_G(name_msg_cache), k2, strlen(k2), (void*)m);
+    free(k2);
+  }
```
done

bshaffer commented on php/ext/google/protobuf/names.c:
```diff
@@ -234,12 +252,43 @@ char *GetPhpClassname(const upb_FileDef *file, const char *fullname) {
+  while(lower[i]) {
+    lower[i] = nolocale_tolower(lower[i]);
+    i++;
+  }
```
done

bshaffer commented on php/ext/google/protobuf/names.c:
```diff
@@ -234,12 +252,43 @@ char *GetPhpClassname(const upb_FileDef *file, const char *fullname) {
+  classname += 2;
+  int length = strlen(classname);
+  char* lower = calloc(1, length + 1);
+  memcpy(lower, classname, length);
```
done

bshaffer commented on php/ext/google/protobuf/names.c:
```diff
@@ -99,6 +102,16 @@ bool is_reserved_name(const char* name) {
 
+bool is_previously_unreserved_name(const char* name) {
+  int i;
+  for (i = 0; kPreviouslyUnreservedNames[i]; i++) {
```
done
